zkevm-circuits/src/pi_circuit2.rs

```diff
@@ -380,1147 +220,479 @@ impl<F: Field> SubCircuitConfig<F> for PiCircuitConfig<F> {
-                region.constrain_equal(txs_hash_hi_cell.cell(), origin_txs_hash_hi_cell.cell())?;
-                region.constrain_equal(txs_hash_lo_cell.cell(), origin_txs_hash_lo_cell.cell())?;
-                Ok((
-                    [block_hash_hi_cell, block_hash_lo_cell],
-                    txs_rlc_acc,
-                    block_rlc_acc,
-                ))
+
+                // input_rlc in self.rpi
+                // input_len in self.rpi_len_acc
+                // output_rlc in self.rpi_rlc_acc
+                let keccak_row = offset;
+                let rpi_rlc_acc_cell = rpi_rlc_acc_cell.unwrap();
+                rpi_rlc_acc_cell.copy_advice(
+                    || "keccak(rpi)_input",
+                    region,
+                    self.rpi,
+                    keccak_row,
+                )?;
+                let keccak = public_data.get_pi();
+                let keccak_rlc =
+                    keccak
+                        .to_fixed_bytes()
+                        .iter()
+                        .fold(Value::known(F::zero()), |acc, byte| {
+                            acc.zip(challenges.keccak_input())
+                                .and_then(|(acc, randomness)| {
+                                    Value::known(acc * randomness + F::from(*byte as u64))
+                                })
+                        });
+                region.assign_advice(
+                    || "rpi_len_acc",
+                    self.rpi_len_acc,
+                    keccak_row,
+                    || Value::known(F::from(rpi_len_acc)),
+                )?;
+                let keccak_output_cell = region.assign_advice(
+                    || "keccak(rpi)_output",
+                    self.rpi_rlc_acc,
+                    keccak_row,
+                    || keccak_rlc,
+                )?;
+                self.q_keccak.enable(region, keccak_row)?;
+
+                rpi_rlc_acc = Value::known(F::zero());
+                offset += 1;
+                // the high 16 bytes of keccak output
+                let mut cells = self.assign_pi_field(
+                    region,
+                    &mut offset,
+                    "high_16_bytes_of_keccak_rpi",
+                    &keccak.to_fixed_bytes()[..16],
+                    &mut rpi_rlc_acc,
+                    &mut rpi_len_acc,
+                    challenges,
+                    true,
+                )?;
+                let keccak_hi_cell = cells[RPI_CELL_IDX].clone();
+
+                // the low 16 bytes of keccak output
+                cells = self.assign_pi_field(
+                    region,
+                    &mut offset,
+                    "low_16_bytes_of_keccak_rpi",
+                    &keccak.to_fixed_bytes()[16..],
+                    &mut rpi_rlc_acc,
+                    &mut rpi_len_acc,
+                    challenges,
+                    true,
+                )?;
+                let keccak_lo_cell = cells[RPI_CELL_IDX].clone();
```

Comment:
Should be possible to make a helper function/loop so more code can be shared because both are identical except for the range.